In [37]:
%load_ext autoreload
%autoreload

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [38]:
import warnings
import os
from pathlib import Path
import pandas as pd
from dotenv import load_dotenv
from analysis.risk_hypthesis_tester import RiskHypothesisTester
from utils.load_data import load_local_data

In [39]:
#suppress warning errors
warnings.filterwarnings("ignore")
#load enviroment variables
load_dotenv()

True

In [40]:
# Automatically go to project root (where .git or README.md is)
project_root = Path.cwd()
while not (project_root / "README.md").exists() and project_root != project_root.parent:
    project_root = project_root.parent

os.chdir(project_root)
print("Project root set to:", project_root)


Project root set to: /home/teshager/Documents/10Academy/repositories/projects/insurance-risk-modeling


## 🧱 1. Setup & Imports

In [41]:
# Load the insurance data
processed_data_dir=os.getenv("PROCESSED_DATA")
file_path= os.path.join(processed_data_dir,'cleaned_insurance_data.csv')
df=pd.read_csv(file_path).drop('CustomValueEstimate',axis=1)
df.head()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title                 Bank      AccountType  \
0              Close Corporation    Mr  First National Bank  Current account   
1              Close Corporation    Mr  First National Bank  Current account   
2              Close Corporation    Mr  First National Bank  Current account   
3              Close Corporation    Mr  First National Bank  Current account   
4              Close Corporation    Mr  First National Bank  Current account   

   MaritalStatus  ... CalculatedPremiumPerTerm  \
0  Not specified  ...                  25.0000   
1  Not specified  ...                  25.0000   
2  Not specified  ...                  25.0000   
3  Not specified  ...                 584.6468   
4  Not specified  ...                 584.6468   

                     ExcessSelected CoverCategory   CoverType  \
0             Mobility - Windscreen    Windscreen  Windscreen   
1             Mobility - Windscreen    Windscreen  Windscreen   
2             Mobility - Windscreen    Windscreen  Windscreen   
3  Mobility - Metered Taxis - R2000    Own damage  Own Damage   
4  Mobility - Metered Taxis - R2000    Own damage  Own Damage   

             CoverGroup              Section                          Product  \
0  Comprehensive - Taxi  Motor Comprehensive  Mobility Metered Taxis: Monthly   
1  Comprehensive - Taxi  Motor Comprehensive  Mobility Metered Taxis: Monthly   
2  Comprehensive - Taxi  Motor Comprehensive  Mobility Metered Taxis: Monthly   
3  Comprehensive - Taxi  Motor Comprehensive  Mobility Metered Taxis: Monthly   
4  Comprehensive - Taxi  Motor Comprehensive  Mobility Metered Taxis: Monthly   

  TotalPremium TotalClaims  Gender_Inferred  
0    21.929825         0.0             Male  
1    21.929825         0.0             Male  
2     0.000000         0.0             Male  
3   512.848070         0.0             Male  
4     0.000000         0.0             Male  

[5 rows x 45 columns]

In [42]:
df.columns.sort_values()

Index(['AccountType', 'AlarmImmobiliser', 'Bank', 'CalculatedPremiumPerTerm',
       'CapitalOutstanding', 'Citizenship', 'Converted', 'CoverCategory',
       'CoverGroup', 'CoverType', 'CrossBorder', 'Cylinders', 'ExcessSelected',
       'Gender_Inferred', 'IsVATRegistered', 'LegalType', 'MainCrestaZone',
       'MaritalStatus', 'Model', 'NewVehicle', 'NumberOfDoors', 'PolicyID',
       'PostalCode', 'Product', 'Province', 'Rebuilt', 'RegistrationYear',
       'Section', 'SubCrestaZone', 'SumInsured', 'TermFrequency', 'Title',
       'TotalClaims', 'TotalPremium', 'TrackingDevice', 'TransactionMonth',
       'UnderwrittenCoverID', 'VehicleIntroDate', 'VehicleType', 'WrittenOff',
       'bodytype', 'cubiccapacity', 'kilowatts', 'make', 'mmcode'],
      dtype='object')

In [43]:
# Define columns
group_col = 'Gender'
metric_col = 'TotalClaims'

#rename the gender_inferred column

df=df.rename(columns={'Gender_Inferred':"Gender"})
df.columns


Index(['UnderwrittenCoverID', 'PolicyID', 'TransactionMonth',
       'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Bank',
       'AccountType', 'MaritalStatus', 'Province', 'PostalCode',
       'MainCrestaZone', 'SubCrestaZone', 'mmcode', 'VehicleType',
       'RegistrationYear', 'make', 'Model', 'Cylinders', 'cubiccapacity',
       'kilowatts', 'bodytype', 'NumberOfDoors', 'VehicleIntroDate',
       'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding',
       'NewVehicle', 'WrittenOff', 'Rebuilt', 'Converted', 'CrossBorder',
       'SumInsured', 'TermFrequency', 'CalculatedPremiumPerTerm',
       'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section',
       'Product', 'TotalPremium', 'TotalClaims', 'Gender'],
      dtype='object')

In [44]:

# Initialize the tester
tester = RiskHypothesisTester(df)

In [45]:
df['Gender'].value_counts()

Gender
Male      933812
Female     65732
Name: count, dtype: int64

In [46]:
# Run the test
# result = tester.two_sample_ttest(group_col=group_col, metric_col=metric_col, equal_var=False)
result= tester.run_ttest(feature=group_col,metric=metric_col)

[2025-06-13 14:13:42,850] INFO: Running t-test for TotalClaims by Gender...
[2025-06-13 14:13:42,850] INFO: Running t-test for TotalClaims by Gender...
[2025-06-13 14:13:42,850] INFO: Running t-test for TotalClaims by Gender...
INFO:analysis.risk_hypthesis_tester:Running t-test for TotalClaims by Gender...


  T-Test Result: Gender vs TotalClaims  
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Metric      ┃ Value                  ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ Group A     │ Male                   │
│ Group B     │ Female                 │
│ T-Statistic │ 2.2504                 │
│ P-Value     │ 0.0244                 │
│ Conclusion  │ REJECT null hypothesis │
└─────────────┴────────────────────────┘

In [47]:
# View result
print(result)

{'feature': 'Gender', 'metric': 'TotalClaims', 'group_a': 'Male', 'group_b': 'Female', 't_stat': 2.250445015518036, 'p_value': 0.024423519322613906, 'conclusion': 'REJECT'}


In [48]:
tester.interpret_result(result)

'We reject the null hypothesis that `Gender` has no effect on `TotalClaims` (p = 0.0244). This suggests that `Gender` is a significant factor affecting `TotalClaims`. Consider using it in segmentation strategy or premium pricing.'

<div style="border-radius:10px; border:orange solid; padding: 15px; font-size:100%; text-align:left; font-color:#325939;background-color:#2c2c2c">
<h3 align="left"><font color='orange'>💡 Observations:</font></h3>

* 'We reject the null hypothesis that `Gender` has no effect on `TotalClaims` (p = 0.0244). This suggests that `Gender` is a significant factor affecting `TotalClaims`. Consider using it in segmentation strategy or premium pricing.'
</div>

In [49]:
tester.export_markdown_report(result)

[2025-06-13 14:16:19,472] INFO: Markdown report saved to hypothesis_report.md
[2025-06-13 14:16:19,472] INFO: Markdown report saved to hypothesis_report.md
[2025-06-13 14:16:19,472] INFO: Markdown report saved to hypothesis_report.md
INFO:analysis.risk_hypthesis_tester:Markdown report saved to hypothesis_report.md
